### PR TITLE
FQDN validation

### PR DESCRIPTION
**Pull request by arturo-seijas**
Applicable spec: N/A

### Overview

FQDN validation for the requests

### Rationale

N/A

### Module Changes

forms.py

### Checklist

- [ ] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [ ] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

**Review by merkata — APPROVED** [review]
nice

File: httprequest_lego_provider/forms.py [hunk 1]
```diff
@@ -2,8 +2,66 @@
...
+    Args:
+        fqdn: the FQDN to validate.
+
+    Returns:
+        if the FQDN is valid.
+    """
+    return fqdn.startswith(FQDN_PREFIX) and _is_fqdn(fqdn.split(".", 1)[1])
+
+
+class FQDNField(CharField):
```
Comment (Thanhphan1147) [inline httprequest_lego_provider/forms.py:46]: I have just a small question, is there any reason why we use this and not pydantic?

File: httprequest_lego_provider/forms.py [hunk 2]
```diff
@@ -2,8 +2,66 @@
...
+                r"^(?!.{255}|.{253}[^.])([a-z0-9](?:[-a-z-0-9]{0,61}[a-z0-9])?\.)+"
+                r"[a-z0-9](?:[-a-z0-9]{0,61}[a-z0-9])?[.]?$"
+            ),
+            fqdn,
+            re.IGNORECASE,
+        )
+    )
+
+
+def is_fqdn_compliant(fqdn) -> bool:
```
Comment (yanksyoon) [inline httprequest_lego_provider/forms.py:34]: I think an input type annotation here can help :D

File: httprequest_lego_provider/tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+from django.contrib.auth.models import User
+
+from httprequest_lego_provider.forms import FQDN_PREFIX
+from httprequest_lego_provider.models import Domain, DomainUserPermission
+
+
+@pytest.fixture(scope="module")
+def username():
+    """Provide a default username."""
+    yield "test_user"
```
Comment (yanksyoon) [inline httprequest_lego_provider/tests/conftest.py:19]: May I ask why `yield` is used here rather than `return`? Would this not create multiple contexts?

**Review by yanksyoon — APPROVED** [review]
Minor questions, LGTM!

File: httprequest_lego_provider/forms.py (hunk 1 above)
Comment (arturo-seijas) [reply httprequest_lego_provider/forms.py:46]: This mechanism is already provided by django, so why adding additional deps?

File: httprequest_lego_provider/tests/conftest.py (hunk 3 above)
Comment (arturo-seijas) [reply httprequest_lego_provider/tests/conftest.py:19]: AFAIK there isn't any difference in this case

File: httprequest_lego_provider/forms.py (hunk 2 above)
Comment (arturo-seijas) [reply httprequest_lego_provider/forms.py:34]: Will add in the next PR. Sorry

**Pull request merged**